### DISCUSSION TITLE
How do I tell Ansible my collection does not support Python < 3.6?

### DISCUSSION THREAD

**Discussion by SHxKM** [ansible-collections/overview#189] [Q&A] [answered]
We're currently working on a new collection that we'd eventually like to have included in `ansible`. We plan to host this collection in our own repo, and utilize the [suggested Github action](https://github.com/ansible-collections/ansible.utils/actions/runs/516050909/workflow) for sanity/unit tests. I'm not sure about a few things though:

1. Are there any tests that are Python 2 specific?
2. I assume that once we [submit an inclusion request](https://github.com/ansible-collections/ansible-inclusion), at some point in ansible's deployment pipeline, the sanity/unit tests are going to be ran against our collection; is there a need to specify somewhere (except documentation) that the collection shouldn't support Python < 3.6?

**Comment by felixfontein** [comment, marked as answer]
Right now there isn't much except the documentation. For ansible-test included in ansible-core 2.12+, you can provide a config file: https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/config/config.yml Simply place it in the tests/ directory of your collection.

If you plan to go through the inclusion process, you want to take a look at https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#python-compatibility which explains what we expect to be documented where.

**Comment by SHxKM** [comment]
Thanks @felixfontein! I've already went through that document (many times) and still implementing all the requirements, I have a follow-up on this snippet:

> For example, if your collection supports Ansible 2.9 to ansible-core 2.13, the Python versions supported for modules are 2.6, 2.7, and 3.5 and newer (until at least 3.10), while the Python versions supported for plugins are 2.7 and 3.5 and newer (until at least 3.10). So if the modules in your collection do not support Python 2.6, you have to document this in the README, for example The content in this collection supports Python 2.7, Python 3.5 and newer..

Does that mean that in order to support `Ansible 2.9`, I need to also support `Python 2.7`? We simply cannot support `Python 2.7` and I'm wondering about the various places that should be configured in..

**Reply by felixfontein** [reply]
No. It means that if your collection supports Ansible 2.9, but does not support Python 2.7 (or 2.6), you have to document it and you need a better reason than "we just don't like Python 2.7" ;-) ("Dependency X we are relying on does not support Python 2.7" is a valid reason.)

**Reply by tadeboro** [reply]
> Does that mean that in order to support `Ansible 2.9`, I need to also support `Python 2.7`? We simply cannot support `Python 2.7` and I'm wondering about the various places that should be configured in.

No, you do not have to support it. But if you do not, you must document this in the README file and possibly in each module. FOr example, this is what we have in our Sensu collection: https://github.com/sensu/sensu-go-ansible/blob/master/plugins/doc_fragments/requirements.py All modules then include this fragment which results in the documentation block at https://docs.ansible.com/ansible/latest/collections/sensu/sensu_go/asset_module.html#requirements

**Comment by tadeboro** [comment]
Hi @SHxKM 

> * Are there any tests that are Python 2 specific?

In general, no. There are some sanity checks that only run on Python 3, but you do not have to worry about those. `ansible-test` will do the right thing for you when running sanity tests.

> * I assume that once we [submit an inclusion request](https://github.com/ansible-collections/ansible-inclusion), at some point in ansible's deployment pipeline, the sanity/unit tests are going to be ran against our collection; is there a need to specify somewhere (except documentation) that the collection shouldn't support Python < 3.6?

The main thing you should do is make sure your sanity test pass. This means that running `ansible-test sanity --docker` passes without errors on all ansible-core versions your collection supports. Currently, you should test at least 2.11 and 2.12. Testing against the development version is also a good idea.

For ansible-core < 2.12, you will need to skip/ignore some tests in the `tests/sanity/ignore-X.Y.txt`. See https://github.com/sensu/sensu-go-ansible/blob/b25e63c945a922ce643f60ff6cae435600a3cf52/tests/sanity/ignore-2.10.txt for some examples. For 2.12, you can configure supported python versions in a single place like we did here: https://github.com/sensu/sensu-go-ansible/blob/master/tests/config.yml

**Reply by SHxKM** [reply]
Thanks @tadeboro! Is there anyway I can test vs different Ansible versions **locally**? I know I can do:

```shell
ansible-test sanity --docker default -v --python 3.6
```

To test against different Python versions, but If I'm not mistaken this is only testing against my local Ansible version...any insights?

**Reply by felixfontein** [reply]
ansible-test is part of the ansible-core/-base/Ansible 2.9 installation, so you can install different versions of ansible-core/ansible-base/Ansible 2.9 in venvs and run ansible-test from these.

**Reply by SHxKM** [reply]
`2.9` doesn't have `ansible_collections` though, does it?

This is what I get when on a venv with Ansible 2.9, and trying to run test-sanity:

```
ERROR: The current working directory must be at or below:

 - an Ansible collection: {...}/ansible_collections/{namespace}/{collection}/
```

**Reply by tadeboro** [reply]
@SHxKM You need to place your collection into properly-namespaced directory structure. See https://sensu.github.io/sensu-go-ansible/hacking/setup.html for instructions we prepared for Sensu and adjust them for your own collection.

**Reply by SHxKM** [reply]
@tadeboro  I've already done that and successfully ran `ansible-test sanity --docker default -v --python 3.6` (for example) with Ansible 2.11, etc..however, Ansible 2.9 seems to be a bit different? when I enter its directory in the venv it's installed in, there is no collections directory.

**Reply by tadeboro** [reply]
Ansible 2.9 is the last monolithic release before the introduction of collections, so yes, the modules in that release are not partitioned into collections. But that version should run tests on collections just fine (we do run them daily IIRC).

**Reply by SHxKM** [reply]
>  But that version should run tests on collections just fine (we do run them daily IIRC).

But how can I do that locally if I can't possibly be in a `/collections` context?

**Reply by felixfontein** [reply]
@SHxKM I'm not sure what you mean with "`/collections` context"? Ansible 2.9 supports collections essentially the same way as ansible-base 2.10 and ansible-core 2.11 do (except that there is no plugin routing functionality, i.e. `meta/runtime.yml` is not supported).

**Reply by SHxKM** [reply]
Ok. Apologies. Let’s try again: I want to test my collection against Ansible 2.9 *locally*, what are the steps I need to take?

**Reply by felixfontein** [reply]
Install Ansible 2.9, and do the same you would do for testing with ansible-base 2.10 or ansible-core 2.11: run `ansible-test <arguments>` inside your collection root directory, which needs to be in the correct path structure (same as for 2.10 and 2.11).

**Reply by SHxKM** [reply]
And simply being inside the 2.9/2.10/2.11 venv dictates which Ansible version the tests are ran against?

**Reply by felixfontein** [reply]
Yes.

**Comment by SHxKM** [comment]
Thanks so much @tadeboro and @felixfontein! It's amazing the quality of help developers get around here!! I'll see you very soon because apart from the new collection we have some complicated deprecations to do in `community.general`. Thanks again!
